### PR TITLE
Alternative approach to Service Workers

### PR DESCRIPTION
**Pull request by ZauberNerd**
This is an alternative approach to #375

### CODE REVIEW

File: packages/build-config/sections/module-rules/webmanifest.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+var hopsConfig = require('hops-config');
+
+exports.default = {
+  test: /\.webmanifest$/,
+  use: [
+    {
+      loader: require.resolve('file-loader'),
+      options: {
+        name: path.join(hopsConfig.assetPath, '[name]-[hash:16].[ext]'),
```
Comment (ZauberNerd) [inline packages/build-config/sections/module-rules/webmanifest.js]: Should we add `emitFile: false` here too?

File: packages/build-config/shims/worker-shim.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+'use strict';
+
+require('babel-polyfill');
+
+(function execute() {
+  var entryPoint = require('hops-worker-entry-point');
+  if (entryPoint.__esModule) {
```
Comment (ZauberNerd) [inline packages/build-config/shims/worker-shim.js]: Should probably also be `typeof entryPoint !== 'function' && typeof entryPoint.default === 'function'`

File: packages/build-config/shims/worker-shim.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    entryPoint = entryPoint.default;
+  }
+  entryPoint(HOPS_ASSETS); // eslint-disable-line no-undef
+  if (module.hot) {
```
Comment (ZauberNerd) [inline packages/build-config/shims/worker-shim.js]: Not sure how Service Workers behave with hot module reloading.. needs more investigation.

File: packages/pwa/dom.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+module.exports = function installServiceWorker() {
+  return new Promise(function(resolve, reject) {
+    if (!('serviceWorker' in navigator)) {
+      return reject(
+        new Error('ServiceWorkers are not supported in this browser')
+      );
+    }
+
+    if (
+      window.location.protocol === 'https' ||
```
Comment (ZauberNerd) [inline packages/pwa/dom.js]: Should we have Service Workers in development and production builds? Or do we need to treat these differently?

File: packages/pwa/dom.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+      window.location.hostname === 'localhost'
+    ) {
+      window.addEventListener('load', function() {
+        resolve(
+          navigator.serviceWorker.register(hopsConfig.workerPath, {
+            scope: hopsConfig.basePath,
```
Comment (ZauberNerd) [inline packages/pwa/dom.js]: Should we always use the `basePath` as scope or are there cases, where we would like to specify a scope that is different from the `basePath`?

File: packages/pwa/node.js [hunk 6]
```diff
@@ -0,0 +1,5 @@
+'use strict';
+
+module.exports = function installServiceWorke() {
+  return new Promise(function() {});
```
Comment (ZauberNerd) [inline packages/pwa/node.js]: Should add a comment to explain that this promise intentionally never resolves. (Can't register a Service Worker in Node.js environment).

File: packages/build-config/plugins/webpack-service-worker.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            compiler.context,
+            require.resolve('../shims/worker-shim'),
+            'worker'
+          ),
+        ])
+        .runAsChild(callback);
+    }
+
+    function onEmit(compilation, callback) {
+      compilation.assets[assetPath] = new RawSource(
```
Comment (ZauberNerd) [inline packages/build-config/plugins/webpack-service-worker.js]: Maybe this should also contain `locations`? Or we fetch `locations` inside Service Worker directly by importing `hopsConfig.locations`.

File: packages/pwa/dom.js (hunk 5 above)
Comment (dmbch) [reply packages/pwa/dom.js]: As `basePath` is the actual application's root path, I think we do not have to cater for this.

File: packages/build-config/sections/module-rules/webmanifest.js (hunk 1 above)
Comment (dmbch) [reply packages/build-config/sections/module-rules/webmanifest.js]: Yes, please. And in node, we could probably also get rid of the `web-app-manifest-loader` itself.

File: packages/build-config/plugins/webpack-service-worker.js (hunk 7 above)
Comment (robin-drexler) [reply packages/build-config/plugins/webpack-service-worker.js]: +1 for hopsConfig. It's how you'd access locations anywhere else too, so I guess it make sense to have it consistent.

File: packages/build-config/shims/worker-shim.js (hunk 3 above)
Comment (robin-drexler) [reply packages/build-config/shims/worker-shim.js]: Shall we disable it for now?

File: packages/config/index.js [hunk 8]
```diff
@@ -29,6 +29,8 @@ function applyDefaultConfig(config) {
       locations: [],
       basePath: '',
       assetPath: '',
+      workerPath: '<assetPath>/sw.js',
```
Comment (dmbch) [inline packages/config/index.js]: please replace `<assetPath>` with `<basePath>` to make sure the SW can control the actual app

File: packages/pwa/dom.js (hunk 4 above)
Comment (dmbch) [reply packages/pwa/dom.js]: We _do_ support SSL in both prod and dev environments, so I would prefer keeping things simple, here.

**Pull request merged**